Hook up a basic foreign-language test harness

rfk opened the pull request:
This is a pitch for how we might do #5, turning our little suite of example foreign-language scripts into actual testcases. It works well enough for me to show it for some feedback, but if we like it, it still needs a bit of polish before merge.

Broad strokes of the PR:

* The `uniffi` crate has been moved into a subdirectory, and it is now part of a single cargo workspace along with the example crates.
* There is a new module `uniffi::support::tests` which builds on the existing ability to run foreign language scripts via an interpreter, and exposes some methods for executing a given script file as a testcase. There is some refactoring of existing modules to make this work nicely, including a significant cleanup of the `uniffi::bindings::*` modules for consistency.
* There is a new crate `uniffi_macros` that exports a single macro, `build_foreign_language_testcases!()`. The idea is that you give it a list of files and it turns them into a suite of `#[test]` functions, one per file.
  * To be honest this probably doesn't need to be a macro, or at least not a procedural macro. But I wanted a small excuse to play around with procedural macros, and I quite like the result.
* The example scripts from `./examples/arithmetic` have been converted into concrete testcases. Not that they really test very much, but they should give you an idea of the resulting consumer experience.
  * If we like the shape of this, I'll have a go at converting the other example crates as well.
* As a side-effect of all this shuffling around, I've also addressed some clippy lints and general code cleanup (to give that satisfying clean `cargo build`-without-warnings feeling).

All of this is geared around a single high-level goal: make `cargo test` in the top-level directory behave in a useful and sensible manner. Doing `cargo test` runs the full test suite for all examples in all target languages. Doing `cargo test py` runs only the python-related tests. And so-on.

Thoughts @linacambridge @jhugman @thomcc?

---

TODO before merge:

* [x] convert the remaining example scripts into testcases.
* [x] update various readmes/docs to mention the new structure.

rfk commented on uniffi/src/bindings/kotlin/mod.rs:
```diff
@@ -56,10 +48,8 @@ pub fn compile_kotlin_bindings(ci: &ComponentInterface, out_dir: &OsStr) -> Resu
-        .wait()
-        .map_err(|_| anyhow::anyhow!("failed to wait for `kotlinc` subprocess"))?;
+        .spawn()?
+        .wait()?;
```
Turns out I had a while bunch of spurious `map_err` throughout the codebase; `std::io::Error` is `Into<anyhow::Error>` so they're not needed.

linacambridge replied:
I’m wondering if this will still give us nice context for errors—“failed to wait for `kotlinc` subprocess” pinpoints the problem much better than some generic “process exited with status code 1337”. Maybe Anyhow’s [`context` and `with_context`](https://docs.rs/anyhow/1.0.13/anyhow/trait.Context.html) are the ticket there. Not saying we need to sprinkle those everywhere, just that I liked your old error messages too! 😸

rfk commented on uniffi/src/bindings/swift/mod.rs:
```diff
@@ -112,68 +106,43 @@ pub fn compile_swift_module(ci: &ComponentInterface, out_dir: &OsStr) -> Result<
-                    // don't exist. Gross!
-                    return Vec::new();
+        cmd.arg("-I").arg(out_dir).arg("-L").arg(out_dir);
+        for entry in PathBuf::from(out_dir).read_dir()? {
```
@linacambridge here I've tried to address https://github.com/rfk/uniffi-rs/issues/16 by looking for files in the target directory, instead of hard-coding them. It seems to work in my local testing. WDYT?

rfk commented on uniffi/src/lib.rs:
````diff
@@ -27,7 +27,7 @@
 //! For example you might define an interface like this:
 //!
-//! ```
+//! ```text
````
Otherwise `cargo test` tries to run these as doctests by default, which doesn't end well.

rfk commented on uniffi/src/lib.rs:
```diff
@@ -258,110 +262,64 @@ fn run_bindings_generate_subcommand(default_lib_file: Option<std::ffi::OsString>
+                Some(target_dir) => {
+                    let ci = bindings::get_component_interface_from_cdylib(&lib_file)?;
+                    bindings::write_bindings(&ci, target_dir, lang)?;
+                    bindings::compile_bindings(&ci, target_dir, lang)?;
```
This now re-generates and re-compiles the bindings whenever you `cargo run -- exec` a script, but I think that's an improvement in ergonomics overall, as you don't have to remember to do that by hand.

linacambridge replied:
👍 When I was working on Swift, I’d forget to do `cargo run -- generate` before `exec` all the time. Definitely an improvement!

linacambridge commented on uniffi/src/bindings/kotlin/mod.rs:
```diff
@@ -69,40 +59,32 @@ pub fn compile_kotlin_bindings(ci: &ComponentInterface, out_dir: &OsStr) -> Resu
+            if let Some(ext) = entry.path().extension() {
+                if ext == "jar" {
+                    classpath.push(":");
+                    classpath.push(entry.path().to_str().unwrap());
```
I think `entry.path().as_os_str()` will give you an `OsStr` that you can append without converting and unwrapping.

rfk replied:
Thanks; turns out plain old `entry.path()` also works

linacambridge commented on uniffi/src/bindings/mod.rs:
```diff
@@ -0,0 +1,138 @@
+/// The functions in this module will delegate to a language-specific backend based
+/// on the provided `TargetLanguage`. For convenience of calling code we also provide
+/// a few `TryFrom` implementations to help guess the correct target language from
+/// e.g. a file extension of command-line argument.
```
Nice refactoring! 🚀

linacambridge commented on uniffi/src/support/tests.rs:
```diff
@@ -0,0 +1,135 @@
+    let language: bindings::TargetLanguage = language.try_into()?;
+    // Have we already generated these bindings?
+    let mut generated_bindings = GENERATED_BINDINGS.lock().unwrap();
+    // XXX TODO: figure out correct type of `HashSet` so that I don't need to `clone()` here...
```
Ooh, good question! I wonder if this will need a tuple struct with a custom `Borrow<&(&PathBuf, &PathBuf, &TargetLanguage)>` implementation, or if you can sprinkle the right combination of `&`s  that’ll make the compiler understand this...

linacambridge commented on uniffi_macros/src/lib.rs:
```diff
@@ -0,0 +1,76 @@
+            );
+            quote! {
+                #[test]
+                fn #test_name () {
```
Could this return a `Result` instead of unwrapping to retain the stack trace if the test case throws?

linacambridge commented on uniffi_macros/src/lib.rs:
```diff
@@ -0,0 +1,76 @@
+/// This macro provides some plumbing to write automated tests for the generated
+/// foreign language bindings of a component. As a component author, you can write
+/// script files in the target foreign language(s) that exercise you component API,
+/// and then call this macro to produce a `cargo test` testcase from each one.
```
Oooh, this is exciting!

rfk commented on uniffi/src/bindings/swift/mod.rs:
```diff
@@ -112,68 +106,43 @@ pub fn compile_swift_module(ci: &ComponentInterface, out_dir: &OsStr) -> Result<
+                    cmd.arg(option);
+                } else if ext == "dylib" || ext == "so" {
+                    let mut option = OsString::from("-l");
+                    option.push(entry.path());
```
I didn't realize it either, I accidentally tried it and it appeared to work so... 🤷‍♂️

rfk commented:
Note to self: the change to make this all compile in a single workspace, doesn't play well with swift bridging headers; building multiple components give me this error: 

```
/Users/rfk/repos/rfk/uniffi-rs/target/debug/uniffi_sprites.modulemap:4:21: error: umbrella for module 'uniffi_geometry' already covers this directory
    umbrella header "/Users/rfk/repos/rfk/uniffi-rs/target/debug/sprites-Bridging-Header.h"
```

rfk commented:
> the change to make this all compile in a single workspace, doesn't play well with swift bridging headers

I fixed this with a bit of subdirectory chicanery, it's not pretty but it seems to work.

rfk commented on .circleci/config.yml:
```diff
@@ -22,7 +24,9 @@ jobs:
       - image: rfkelly/uniffi-ci:latest
+    resource_class: small
     steps:
+      - run: cat ~/.profile >> $BASH_ENV
```
I added a `.circleci/config.yaml` directly on master so that circle would start building things, but I'd welcome any feedback on it as part of this PR.

rfk commented on docker/Dockerfile-build:
```diff
@@ -57,4 +57,14 @@ RUN mkdir -p /tmp/setup-kotlin \
+    && echo "export CLASSPATH=\"\$CLASSPATH:/opt/jna.jar\"" >> /home/circleci/.bashrc \
+    && echo "export CLASSPATH=\"\$CLASSPATH:/opt/jna.jar\"" >> /home/circleci/.profile \
+    && cd ../ \
+    && rm -rf ./setup-jna
```
I also added a dockerfile that builds all the tools we need, but I forgot to install JNA on my first attempt, so here it is as part of this PR :-)

If you're curious about the whole file, you can view it [here](https://github.com/rfk/uniffi-rs/blob/becd87c30bceca0af989266ca75633cc7059fd6f/docker/Dockerfile-build).

rfk commented on examples/naming-conventions/Cargo.toml:
```diff
@@ -1,19 +0,0 @@
-[package]
-name = "uniffi-example-naming-conventions"
```
Perhaps controversially, I have deleted the `naming-conventions` example. Reason being that, when I added these tests, I discovered that the `sprites` example was failing in Kotlin due to not being updated for the naming conventions. So I think we can safely check the behaviour of those conventions through the existing examples.

rfk commented:
@tarikeshaq heads-up, if you do end up working on adding string support, you might like to base it on top of this branch in order to get some basic automated test infra.

rfk commented:
Alrighty, I'm happy enough with how this works to take it out of "draft". @linacambridge I'm aware you're out for the next few days so no rush, but please take another look over this at your convenience.

rfk commented on uniffi/src/bindings/swift/mod.rs:
```diff
@@ -0,0 +1,188 @@
+/// in the same workspace.
+///
+/// We work around this by generating the module-mapping instructions in a separate subdirectory
+/// for each module, producing a directory structure that looks like this:
```
@linacambridge in particular could you please take a look at the yucky things I had to do here to make this work in a cargo workspace? You may be able to see a better way.

linacambridge replied:
I think we can actually get rid of the yucky stuff if we change `umbrella header` to just `header` in the `.modulemap` (that's probably technically more correct, too—"umbrella header" means "includes all the other headers in that directory", which our `-Bridging-Header.h` definitely doesn't do. There's probably other details I'm missing here; I just decided to try it on a whim).

I just tried it locally, and both `cargo test` in the top-level workspace, and `cargo run -- exec tests/bindings/test_geometry.swift` worked! 🎉

tarikeshaq commented on uniffi/src/lib.rs:
```diff
@@ -205,11 +206,7 @@ pub fn run_bindgen_helper(component_name: Option<&str>) -> Result<()> {
-                        } else {
-                            true
-                        })
+                        .required(default_lib_file.is_some())
```
```suggestion
                        .required(default_lib_file.is_none())
```

Realized while adding the tests for strings that I had to manually set the lib_file when running `generate` 😄

rfk replied:
nice catch, thanks! 😅

linacambridge commented on uniffi/src/bindings/kotlin/mod.rs:
```diff
@@ -57,9 +48,9 @@ pub fn compile_kotlin_bindings(ci: &ComponentInterface, out_dir: &OsStr) -> Resu
         .arg(jar_file)
         .spawn()
-        .map_err(|_| anyhow::anyhow!("failed to spawn `kotlinc`"))?
+        .context("Failed to `kotlinc` to compile the bindings")?
```
```suggestion
        .context("Failed to spawn `kotlinc` to compile the bindings")?
```

linacambridge commented on uniffi/src/bindings/kotlin/mod.rs:
```diff
@@ -78,30 +69,30 @@ pub fn run_kotlin_script(out_dir: Option<&OsStr>, script_file: Option<&str>) ->
-                        classpath.push(":");
-                        classpath.push(entry.path().to_str().unwrap());
-                    }
+            let entry = entry.context("Directory listing failed while running Koltin script")?;
```
```suggestion
            let entry = entry.context("Directory listing failed while running Kotlin script")?;
```

linacambridge commented on uniffi/src/support/tests.rs:
```diff
@@ -0,0 +1,141 @@
+    let cdylib_file = ensure_compiled_cdylib(pkg_dir)?;
+    let out_dir = Path::new(cdylib_file.as_str())
+        .parent()
+        .ok_or_else(|| anyhow::anyhow!("Generated cdylib has not parent directory"))?;
```
```suggestion
        .ok_or_else(|| anyhow::anyhow!("Generated cdylib has no parent directory"))?;
```

linacambridge commented on uniffi/src/support/tests.rs:
```diff
@@ -0,0 +1,141 @@
+/// Execute the given foreign-language script as part of a rust test suite.
+///
+/// This function takes the top-level directory of a uniffi component crate, and the path to
+/// a foriegn-language test file that exercises that component's bindings. It ensures that the
```
```suggestion
/// a foreign-language test file that exercises that component's bindings. It ensures that the
```

linacambridge commented on uniffi/src/support/tests.rs:
```diff
@@ -0,0 +1,141 @@
+        .filenames
+        .iter()
+        .filter(|nm| match nm.extension().unwrap_or_default().to_str() {
+            Some("dylib") | Some("so") => true,
```
Would this be `dll` for Windows someday?

rfk replied:
Maybe one day, when @mhammond gets sufficiently ~~annoyed~~ motivated to shave that yak 😁

rfk commented:
Thanks @linacambridge, that's much nicer than the gross subdirectory hack! I've updated and squashed and will merge this once CI goes green. (Wow I am relieved to have actual CI here!)
